### ISSUE TITLE
Initial conditions for soil horizons

### ISSUE THREAD

**Issue by schmidtfederico** [RDCEP/psims#5]
Hello,

I'm trying to recreate an experiment a colleague of mine runs in DSSAT for maize crops in Argentina. However, I'm having trouble with the soil initial conditions.

Is it possible to change the soil's initial conditions at each soil horizon? Looking through the `jsons2dssat.py` script I found that the values in that section are derived from the soil.json file in different ways:
- **SH2O**: the script uses a coefficient called `frac_full` and multiplies the maximum amount of water in each horizon by that value (in `soil.json` it's called `sldul` if I'm not mistaken). However, what if there's different levels of moisture at each horizon?
- **SNH4**: the script allows for the modification of this value, however, it uses only the first value (the value at the first soil horizon) and replicates that value for every other horizon. The experiment I'm trying to replicate uses 0.5ppm in the first five soil horizons and 0.2ppm for the others. Is it possible to change this?
- **SNO3**: I couldn't find a way to change this variable, it seems to be an exact copy of the `sloc` variable in the `soil.json`. Am I right?

Here's how the initial conditions look like in the experiment I'm talking about:

```
*INITIAL CONDITIONS
@C   PCR ICDAT  ICRT  ICND  ICRN  ICRE  ICWD ICRES ICREN ICREP ICRIP ICRID  ICNAME
 1    SB   -99  2000     0  1.00  1.00 -99.0  3000   1.5  0.00     0     0
@C  ICBL  SH2O  SNH4  SNO3
 1    15 0.104   0.5  14.0
 1    30 0.229   0.5   7.0
 1    45 0.229   0.5   3.5
 1    70 0.229   0.5   1.8
 1    90 0.207   0.5   0.9
 1   126 0.207   0.2   0.4
 1   150 0.205   0.2   0.2
 1   180 0.205   0.2   0.1
 1   250 0.205   0.2   0.1
```

And here's what I'm getting after running the `jsons2dssat.py` script:

```
*INITIAL CONDITIONS
@C   PCR ICDAT  ICRT  ICND  ICRN  ICRE  ICWD ICRES ICREN ICREP ICRIP ICRID ICNAME
 1    SB  1971  2000     0     2     1   -99  3000  1.00  1.00     0     0 Cond. Name
@C  ICBL  SH2O  SNH4  SNO3
 1    15 0.225   0.5   1.1
 1    30 0.229   0.5   0.8
 1    45 0.229   0.5   0.4
 1    70 0.229   0.5   0.1
 1    90 0.207   0.5   0.1
 1   126 0.207   0.5   0.1
 1   150 0.205   0.5   0.1
 1   180 0.205   0.5   0.1
 1   250 0.205   0.5   0.1
```

Greetings and thanks for your time and effort.
Federico.

**Comment by joshuaelliott** [comment]
hey frederico, 

sorry for the delayed response. i managed to respond to one of your questions and then got buried and stuff and convinced myself i had responded to all of them. 

i thought it was possible to change the initial conditions using the campaign file just as you would any other variable... the initial conditions are in the template so they will be overwritten by camp2json just like anything else if it finds them in the campaign file. 

that said, we do have some automatic functions in some places that will do convenient operations such as setting the fraction of water in each layer to some fixed fraction 'frac_full' or something like that. so if you set the "frac_full parameter in the template file, then that might end up overwriting sh2o paramters that you put in the campaign file. note that with the current methodology if you wanted to set the values for each layer separately in the campaign file, i think you would have to use the _123 notation like you do for "date" variables (because sh2o and snh4 appear more than once). so you would need to set snh4_1, snh4_2, etc... in the campaign file and then the translator will know to only replace the first appearance of snh4 with the snh4_1 value and the second with snh4_2 and etc. currently that's the only good way to do it i'm afraid. 

regarding sno3, we may have set an automatic bevaiour that sets this equal to 1/10th of the organic carbon in order to match standard C/N ratios. because of the units of these values, the numbers end up being the same. i would hope that if you set this variable in the campaign file that this would override this default behaviour, but i'm not totally sure. let me know if it doesn't work as you need and we can try to take a deeper look. do you have a branch of the code somewhere that you are editing?

**Comment by schmidtfederico** [comment]
Dear Joshua,

Thank you for your response. We have managed to get pSIMS up and running since the last time I wrote to you and we have also made a few changes to it.

Regarding your suggestion, though it was possible to modify initial condition variables in the template by using the underscore+index technic, it wasn't enough flexible for us: we are working to build a week-by-week yield forecast in different regions of Argentina and there are different amount of soil horizons at different sites. That approach would require to us to define a fixed amount of layers in the experiment template.
However, after doing some more debugging of jsons2dssat.py I convinced myself that what was happening with soil variables was exactly as I described initially: they were being derived from the soil.json and allowed little modification by using the experiment.json.

Since we considered that pSIMS was the tool we needed to achieve our goal, we decided to build what we needed in it. The changes we made can be found in [schmidtfederico/psims](https://github.com/schmidtfederico/psims) and a detailed description of what changed can be found in the file [CHANGES.md](https://github.com/schmidtfederico/psims/blob/master/CHANGES.md).
The easiest way to compare both repositories would be by using the [Compare Tool](https://github.com/RDCEP/psims/compare/master...schmidtfederico:master?diff=split&name=master) GitHub provides and follow the changes markdown file. However, let me know if there's anything that needs further explanation.

Basically, we created a new dimension inside the NetCDF for soil data called soil_layer. Variables associated with that dimension will be read as arrays and mapped together to build the dictionary of each soil layer.

Finally, we have created a [small working campaign](https://github.com/schmidtfederico/psims/tree/master/campaigns/example/junin) that may help your team to test the new features and also help newcomers to understand and use pSIMS. We've included a Python script that creates the campaign and gives further details on how to define soil variables.

Greetings and thank you for your help.
